### ISSUE TITLE
sql: regclass fails when there are multiple tables with the same name

### ISSUE THREAD

**Issue by mkysel** [MaterializeInc/materialize#9484]
When multiple tables exist with the same name and the regclass cast is not fully qualified, PG will use the search_path to find the table and its OID.

```
martin=# create table testing(i int);
CREATE TABLE
martin=# create table just_for_testing.testing(i int);
CREATE TABLE
martin=# select 'testing'::regclass;
 regclass 
----------
 testing
(1 row)

martin=# select 'testing'::regclass::oid;
  oid  
-------
 27096
(1 row)

martin=# select 'just_for_testing.testing'::regclass::oid;
  oid  
-------
 27099
(1 row)
martin=# show search_path;
   search_path   
-----------------
 "$user", public
(1 row)
martin=# set search_path TO "$user", public,just_for_testing
martin-# ;
SET
martin=# show search_path;
            search_path            
-----------------------------------
 "$user", public, just_for_testing
(1 row)

martin=# select 'testing'::regclass::oid;
  oid  
-------
 27096
(1 row)

martin=# set search_path TO just_for_testing,"$user", public                 
;
SET
martin=# select 'testing'::regclass::oid;
  oid  
-------
 27099
(1 row)
```

vs the behavior in MZ:
```
materialize=> create table s_foo.testing(i int);
CREATE TABLE
materialize=> select 'testing'::regclass;
 ?column? 
----------
 20270
(1 row)

materialize=> create table testing(i int);
CREATE TABLE
materialize=> select 'testing'::regclass;
ERROR:  Evaluation error: more than one record produced in subquery
```

**Comment by jkosh44** [comment]
In general, this function should behave identically to name resolution in `SELECT` statements, instead of just comparing the name to the name column in `mz_objects`.

Some current defects are:
- Qualified names fail to cast properly to `regclass`.
- Objects not in the search path should fail to cast.
- The first matching object in the search_path should be the result of the cast.

The same issue exists for `regproc` and `regtype`.

**Comment by sploiselle** [comment]
Closed by #20407
